<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.345">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Troutman Lab - An introductory exploration of RNA-seq data using R and tidyverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link class="quarto-color-scheme" id="quarto-text-highlighting-styles" href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet">
<link class="quarto-color-scheme quarto-color-alternate" rel="prefetch" id="quarto-text-highlighting-styles" href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link class="quarto-color-scheme" href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<link class="quarto-color-scheme quarto-color-alternate" rel="prefetch" href="../../site_libs/bootstrap/bootstrap-dark.min.css">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Troutman Lab</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html">Home</a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-analysis" role="button" data-bs-toggle="dropdown" aria-expanded="false">Analysis</a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-analysis">    
        <li>
    <a class="dropdown-item" href="../../Pages/Analysis/BMI_Cluster.html">
 <span class="dropdown-text">BMI Cluster</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Pages/Analysis/Lab_Cluster_Drives.html">
 <span class="dropdown-text">Lab Cluster Drives</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Pages/Analysis/processingRNASeqData.html">
 <span class="dropdown-text">Processing RNA-Seq Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://github.com/ty-troutman/Genomics-analyses">
 <span class="dropdown-text">Just beginning?</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-sequencing" role="button" data-bs-toggle="dropdown" aria-expanded="false">Sequencing</a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-sequencing">    
        <li>
    <a class="dropdown-item" href="../../Pages/Sequencing/Database_Barcodes.html">
 <span class="dropdown-text">Database and Barcodes</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Pages/Sequencing/Pooling_Submission.html">
 <span class="dropdown-text">Pooling and Submission</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../Pages/Sequencing/NGSScraper_KevinErnst.html">
 <span class="dropdown-text">Downloading the Data</span></a>
  </li>  
        <li class="dropdown-header">More stuff</li>
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ty-troutman"><i class="bi bi-github" role="img" aria-label="Troutman GitHub">
</i> 
 </a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://www.cincinnatichildrens.org/research/divisions/a/allergy-immunology/labs/troutman">CCHMC</a>
  </li>  
</ul>
              <div class="quarto-toggle-container">
                  <a href="" class="quarto-color-scheme-toggle nav-link" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
              </div>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#some-basics" id="toc-some-basics" class="nav-link active" data-scroll-target="#some-basics">Some basics</a></li>
  <li><a href="#reading-in-data" id="toc-reading-in-data" class="nav-link" data-scroll-target="#reading-in-data">Reading in data</a></li>
  <li><a href="#a-basic-scatter-plot" id="toc-a-basic-scatter-plot" class="nav-link" data-scroll-target="#a-basic-scatter-plot">A Basic Scatter Plot</a></li>
  <li><a href="#finished" id="toc-finished" class="nav-link" data-scroll-target="#finished">Finished!!</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">An introductory exploration of RNA-seq data using R and tidyverse</h1>
</div>





<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<section id="some-basics" class="level1">
<h1>Some basics</h1>
<p>Commenting your code can help you and others understand what you were doing. A comment is generated using <code>#</code>.</p>
<p>You can navigate your computer as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how to find your working directory</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/Users/tro3nr/OneDrive - cchmc/Websites/ty-troutman.github.io/Pages/Sequencing"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how to list directory files</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Bar-Charts_files"                                
 [2] "Bar-Charts.html"                                 
 [3] "Bar-Charts.qmd"                                  
 [4] "Database_Barcodes.html"                          
 [5] "Database_Barcodes.qmd"                           
 [6] "Exporting-Gene-Lists.qmd"                        
 [7] "Integrating-RNASeq-ExpressionTPM-with-DeSeq2.qmd"
 [8] "Introduction-RNASeqAnalysis.qmd"                 
 [9] "Introduction-RNASeqAnalysis.rmarkdown"           
[10] "Nested-Bar-Charts.qmd"                           
[11] "NGSScraper_KevinErnst.qmd"                       
[12] "Pooling_Submission.html"                         
[13] "Pooling_Submission.qmd"                          </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#how to list sub directories</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.dirs</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "."                              "./Bar-Charts_files"            
[3] "./Bar-Charts_files/figure-html"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how to list files from someplace other than your working directory</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"../Data/Source/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "1-s2.0-S107476132030159X-mmc3.xlsx"                 
 [2] "BALBcJ_F0_Control.vs.C57BL6J_F0_Control.scatter.txt"
 [3] "BALBcJ_F1_Control.vs.C57BL6J_F1_Control.scatter.txt"
 [4] "C57BL6J_F0_Control.vs.BALBcJ_F0_Control.scatter.txt"
 [5] "C57BL6J_F1_Control.vs.BALBcJ_F1_Control.scatter.txt"
 [6] "deseq2_kc_control_lps_f0_f1.txt"                    
 [7] "HOMER.rawCount.txt"                                 
 [8] "HOMER.rawTPM.txt"                                   
 [9] "rawC.txt"                                           
[10] "rawT.txt"                                           
[11] "rnaSeq_tableS2_seidman_immunity_2020.csv"           </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how to make a new directory</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># dir.create("~/Documents/example")</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># how to change your working directory</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># setwd("./")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="reading-in-data" class="level1">
<h1>Reading in data</h1>
<p>In this tutorial we will make a scatter plot using RNA-seq data from table S2 in <a href="https://pubmed.ncbi.nlm.nih.gov/32362324/">Seidman et. al.&nbsp;Immunity. Volume 52, Issue 6, 16 June 2020, Pages 1057-1074.e7</a>, which can be downloaded <a href="https://ars.els-cdn.com/content/image/1-s2.0-S107476132030159X-mmc3.xlsx">here</a>. Once we have the data, we will need to open Rstudio and load the necessary packages. We can do that with <a href="https://www.datacamp.com/community/tutorials/r-packages-guide"><code>library()</code></a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✓ ggplot2 3.3.5     ✓ purrr   0.3.4
✓ tibble  3.1.6     ✓ dplyr   1.0.8
✓ tidyr   1.2.0     ✓ stringr 1.4.0
✓ readr   2.1.2     ✓ forcats 0.5.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyr' was built under R version 4.1.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.1.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()</code></pre>
</div>
</div>
<p>Now we can use <code>read_excel()</code> from the <a href="https://readxl.tidyverse.org">readxl</a> package to load our excel spreadsheet. We can use the functions <code>readr:read_csv()</code> or <code>readr:read_tsv()</code> from the <a href="https://readr.tidyverse.org">readr package</a> to load .csv and .txt files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in our data</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span> </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  readxl<span class="sc">::</span><span class="fu">read_excel</span>(</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"../Data/Source/1-s2.0-S107476132030159X-mmc3.xlsx"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>tb1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 48
   `Transcript/RepeatID (cm…` `Annotation/Di…` `KCH_rep1_figu…` `KCH_rep2_figu…`
   &lt;chr&gt;                      &lt;chr&gt;                       &lt;dbl&gt;            &lt;dbl&gt;
 1 NM_001329047               Mia2|Ctage5|D12…            43.6             36.3 
 2 NM_172405                  Abraxas1|383040…             1.91             2.20
 3 NM_001359283               Sec63|5730478J1…            19.3             17.8 
 4 NM_001168290               Sugp2|Sfrs14|Sr…            11.5              8.86
 5 NM_001356498               Mtmr12|3Pap|493…             7.14             8.74
 6 NM_001358950               Cbx5|2610029O15…             6.53             6.21
 7 NM_001166375               March1|2900024D…            58.3             67.8 
 8 NM_178061                  Mob3b|8430436F2…             1.51             1.61
 9 NR_126506                  Mdm4|4933417N07…            29.2             25.6 
10 NR_029557                  Mir145a|Mir145|…             0                0   
# … with 40,448 more rows, and 44 more variables: `KCN_rep1_figure1,3,4` &lt;dbl&gt;,
#   `KCN_rep2_figure1,3,4` &lt;dbl&gt;, `KN-RM_rep1_figure1,3,4` &lt;dbl&gt;,
#   `KN-RM_rep2_figure1,3,4` &lt;dbl&gt;, `Ly6CHi-RM_rep1_figure1,3,4` &lt;dbl&gt;,
#   `Ly6CHi-RM_rep2_figure1,3,4` &lt;dbl&gt;, `Ly6CLow-RM_rep1_figure1,3,4` &lt;dbl&gt;,
#   `Ly6CLow-RM_rep2_figure1,3,4` &lt;dbl&gt;,
#   `BloodLy6CHi-Mono_Control_rep1_figure3,4` &lt;dbl&gt;,
#   `BloodLy6CHi-Mono_Control_rep2_figure3,4` &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>We can use <code>colnames()</code> to see the names of the columns in this table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="at">x =</span> tb1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Transcript/RepeatID (cmd=analyzeRepeats.pl rna mm10 -count exons -tpm -dfile Table_S2_dfile.txt)"
 [2] "Annotation/Divergence"                                                                           
 [3] "KCH_rep1_figure1,3,4"                                                                            
 [4] "KCH_rep2_figure1,3,4"                                                                            
 [5] "KCN_rep1_figure1,3,4"                                                                            
 [6] "KCN_rep2_figure1,3,4"                                                                            
 [7] "KN-RM_rep1_figure1,3,4"                                                                          
 [8] "KN-RM_rep2_figure1,3,4"                                                                          
 [9] "Ly6CHi-RM_rep1_figure1,3,4"                                                                      
[10] "Ly6CHi-RM_rep2_figure1,3,4"                                                                      
[11] "Ly6CLow-RM_rep1_figure1,3,4"                                                                     
[12] "Ly6CLow-RM_rep2_figure1,3,4"                                                                     
[13] "BloodLy6CHi-Mono_Control_rep1_figure3,4"                                                         
[14] "BloodLy6CHi-Mono_Control_rep2_figure3,4"                                                         
[15] "BloodLy6CHi-Mono_NASH_rep1_figure3,4"                                                            
[16] "BloodLy6CHi-Mono_NASH_rep2_figure3,4"                                                            
[17] "BloodLy6CHi-Mono_Control_rep3_figure3,4"                                                         
[18] "BloodLy6CHi-Mono_Control_rep4_figure3,4"                                                         
[19] "RLM_DT1D_rep1_figure3"                                                                           
[20] "RLM_DT1D_rep2_figure3"                                                                           
[21] "RLM_DT1D_rep3_figure3"                                                                           
[22] "RLM_DT14D_rep1_figure3"                                                                          
[23] "RLM_DT14D_rep2_figure3"                                                                          
[24] "RLM_DT14D_rep3_figure3"                                                                          
[25] "RLM_DT12H_rep1_figure3"                                                                          
[26] "RLM_DT12H_rep2_figure3"                                                                          
[27] "wholeLiver_DTRN_rep1_figure3"                                                                    
[28] "wholeLiver_DTRN_rep2_figure3"                                                                    
[29] "wholeLiver_DTRN_rep3_figure3"                                                                    
[30] "wholeLiver_DTRN_rep4_figure3"                                                                    
[31] "wholeLiver_DTRN_rep5_figure3"                                                                    
[32] "wholeLiver_DTRN_rep6_figure3"                                                                    
[33] "wholeLiver_DTRN_rep7_figure3"                                                                    
[34] "wholeLiver_DTRN_rep8_figure3"                                                                    
[35] "wholeLiver_DTRN_rep9_figure3"                                                                    
[36] "wholeLiver_DTRN_rep10_figure3"                                                                   
[37] "wholeLiver_DTRP_rep1_figure3"                                                                    
[38] "wholeLiver_DTRP_rep2_figure3"                                                                    
[39] "wholeLiver_DTRP_rep3_figure3"                                                                    
[40] "wholeLiver_DTRP_rep4_figure3"                                                                    
[41] "wholeLiver_DTRP_rep5_figure3"                                                                    
[42] "wholeLiver_DTRP_rep6_figure3"                                                                    
[43] "wholeLiver_DTRP_rep7_figure3"                                                                    
[44] "wholeLiver_DTRP_rep8_figure3"                                                                    
[45] "wholeLiver_DTRP_rep9_figure3"                                                                    
[46] "wholeLiver_DTRP_rep10_figure3"                                                                   
[47] "wholeLiver_DTRP_rep11_figure3"                                                                   
[48] "wholeLiver_DTRP_rep12_figure3"                                                                   </code></pre>
</div>
</div>
<p>That’s a lot of data, lets only keeps a small part that we want to practice with for now. To do this let’s keep the first two columns, which have gene identifier information, and lets keep the next 4 column which are replicate RNA-seq data for two conditions. To keep these columns, we will use the <a href="https://dplyr.tidyverse.org/reference/select.html"><code>dplyr::select()</code></a> verb, which operates on columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">.data =</span> tb1, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   `Transcript/RepeatID (cm…` `Annotation/Di…` `KCH_rep1_figu…` `KCH_rep2_figu…`
   &lt;chr&gt;                      &lt;chr&gt;                       &lt;dbl&gt;            &lt;dbl&gt;
 1 NM_001329047               Mia2|Ctage5|D12…            43.6             36.3 
 2 NM_172405                  Abraxas1|383040…             1.91             2.20
 3 NM_001359283               Sec63|5730478J1…            19.3             17.8 
 4 NM_001168290               Sugp2|Sfrs14|Sr…            11.5              8.86
 5 NM_001356498               Mtmr12|3Pap|493…             7.14             8.74
 6 NM_001358950               Cbx5|2610029O15…             6.53             6.21
 7 NM_001166375               March1|2900024D…            58.3             67.8 
 8 NM_178061                  Mob3b|8430436F2…             1.51             1.61
 9 NR_126506                  Mdm4|4933417N07…            29.2             25.6 
10 NR_029557                  Mir145a|Mir145|…             0                0   
# … with 40,448 more rows, and 2 more variables: `KCN_rep1_figure1,3,4` &lt;dbl&gt;,
#   `KCN_rep2_figure1,3,4` &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this format gives us column 1 to 6, and everything between</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">.data =</span> tb1, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   `Transcript/RepeatID (cm…` `Annotation/Di…` `KCH_rep1_figu…` `KCH_rep2_figu…`
   &lt;chr&gt;                      &lt;chr&gt;                       &lt;dbl&gt;            &lt;dbl&gt;
 1 NM_001329047               Mia2|Ctage5|D12…            43.6             36.3 
 2 NM_172405                  Abraxas1|383040…             1.91             2.20
 3 NM_001359283               Sec63|5730478J1…            19.3             17.8 
 4 NM_001168290               Sugp2|Sfrs14|Sr…            11.5              8.86
 5 NM_001356498               Mtmr12|3Pap|493…             7.14             8.74
 6 NM_001358950               Cbx5|2610029O15…             6.53             6.21
 7 NM_001166375               March1|2900024D…            58.3             67.8 
 8 NM_178061                  Mob3b|8430436F2…             1.51             1.61
 9 NR_126506                  Mdm4|4933417N07…            29.2             25.6 
10 NR_029557                  Mir145a|Mir145|…             0                0   
# … with 40,448 more rows, and 2 more variables: `KCN_rep1_figure1,3,4` &lt;dbl&gt;,
#   `KCN_rep2_figure1,3,4` &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this format declares which columns to keep in order</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also pipe commands from left to right using magrittr <a href="https://magrittr.tidyverse.org"><code>%\&gt;%</code></a> operator. This allows us to chain operations together so that we don’t need to define a bunch of variables to store in memory, or overwrite variables by reusing variable names. Reusing variable names can lead to bugs in our analyses (if we aren’t careful)!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"../Data/Source/1-s2.0-S107476132030159X-mmc3.xlsx"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>tb1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   `Transcript/RepeatID (cm…` `Annotation/Di…` `KCH_rep1_figu…` `KCH_rep2_figu…`
   &lt;chr&gt;                      &lt;chr&gt;                       &lt;dbl&gt;            &lt;dbl&gt;
 1 NM_001329047               Mia2|Ctage5|D12…            43.6             36.3 
 2 NM_172405                  Abraxas1|383040…             1.91             2.20
 3 NM_001359283               Sec63|5730478J1…            19.3             17.8 
 4 NM_001168290               Sugp2|Sfrs14|Sr…            11.5              8.86
 5 NM_001356498               Mtmr12|3Pap|493…             7.14             8.74
 6 NM_001358950               Cbx5|2610029O15…             6.53             6.21
 7 NM_001166375               March1|2900024D…            58.3             67.8 
 8 NM_178061                  Mob3b|8430436F2…             1.51             1.61
 9 NR_126506                  Mdm4|4933417N07…            29.2             25.6 
10 NR_029557                  Mir145a|Mir145|…             0                0   
# … with 40,448 more rows, and 2 more variables: `KCN_rep1_figure1,3,4` &lt;dbl&gt;,
#   `KCN_rep2_figure1,3,4` &lt;dbl&gt;</code></pre>
</div>
</div>
<p>These column names could be made easier. Let’s rename them using the dplyr verb <a href="https://dplyr.tidyverse.org/reference/rename.html?q=rename"><code>dplyr::rename</code></a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> <span class="st">"../Data/Source/1-s2.0-S107476132030159X-mmc3.xlsx"</span>) </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(tb1, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span> <span class="fu">rename</span>(<span class="at">.data =</span> tb1, <span class="at">accession =</span> <span class="dv">1</span>, <span class="at">gene =</span> <span class="dv">2</span>, <span class="at">kch_rep1 =</span> <span class="dv">3</span>, <span class="at">kch_rep2 =</span> <span class="dv">4</span>, <span class="at">kn_rep1 =</span> <span class="dv">5</span>, <span class="at">kn_rep2 =</span> <span class="dv">6</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>tb1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   accession    gene                           kch_rep1 kch_rep2 kn_rep1 kn_rep2
   &lt;chr&gt;        &lt;chr&gt;                             &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2|Ctage5|D12Bwg0579e|Mea6|…    43.6     36.3   29.6    26.8  
 2 NM_172405    Abraxas1|3830405G04Rik|563040…     1.91     2.20   1.95    1.82 
 3 NM_001359283 Sec63|5730478J10Rik|AI649014|…    19.3     17.8   13.2    12.9  
 4 NM_001168290 Sugp2|Sfrs14|Srsf14|mKIAA0365…    11.5      8.86  10.9     7.57 
 5 NM_001356498 Mtmr12|3Pap|4932703C11|C73001…     7.14     8.74   6.71    6.84 
 6 NM_001358950 Cbx5|2610029O15Rik|C75991|HP1…     6.53     6.21   5.86    7.20 
 7 NM_001166375 March1|2900024D24Rik|BB085186…    58.3     67.8   49.4    50.2  
 8 NM_178061    Mob3b|8430436F23Rik|A430018A0…     1.51     1.61   0.729   0.906
 9 NR_126506    Mdm4|4933417N07Rik|AA414968|A…    29.2     25.6   16.4    13.3  
10 NR_029557    Mir145a|Mir145|Mirn145|mir-14…     0        0      0       0    
# … with 40,448 more rows</code></pre>
</div>
</div>
<p>The above snippet is equivalent to using a string of pipes to parse the data prior to generating the final variable:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="ot">&lt;-</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(<span class="at">path =</span> <span class="st">"../Data/Source/1-s2.0-S107476132030159X-mmc3.xlsx"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">accession =</span> <span class="dv">1</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene =</span> <span class="dv">2</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_rep1 =</span> <span class="dv">3</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_rep2 =</span> <span class="dv">4</span>,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_rep1 =</span> <span class="dv">5</span>,</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_rep2 =</span> <span class="dv">6</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>tb1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   accession    gene                           kch_rep1 kch_rep2 kn_rep1 kn_rep2
   &lt;chr&gt;        &lt;chr&gt;                             &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2|Ctage5|D12Bwg0579e|Mea6|…    43.6     36.3   29.6    26.8  
 2 NM_172405    Abraxas1|3830405G04Rik|563040…     1.91     2.20   1.95    1.82 
 3 NM_001359283 Sec63|5730478J10Rik|AI649014|…    19.3     17.8   13.2    12.9  
 4 NM_001168290 Sugp2|Sfrs14|Srsf14|mKIAA0365…    11.5      8.86  10.9     7.57 
 5 NM_001356498 Mtmr12|3Pap|4932703C11|C73001…     7.14     8.74   6.71    6.84 
 6 NM_001358950 Cbx5|2610029O15Rik|C75991|HP1…     6.53     6.21   5.86    7.20 
 7 NM_001166375 March1|2900024D24Rik|BB085186…    58.3     67.8   49.4    50.2  
 8 NM_178061    Mob3b|8430436F23Rik|A430018A0…     1.51     1.61   0.729   0.906
 9 NR_126506    Mdm4|4933417N07Rik|AA414968|A…    29.2     25.6   16.4    13.3  
10 NR_029557    Mir145a|Mir145|Mirn145|mir-14…     0        0      0       0    
# … with 40,448 more rows</code></pre>
</div>
</div>
</section>
<section id="a-basic-scatter-plot" class="level1">
<h1>A Basic Scatter Plot</h1>
<p>Now we have some tidy data, let’s compare the replicates by making a scatterplot! We can accomplish this using <a href="https://ggplot2.tidyverse.org">ggplot2</a> and the <a href="https://ggplot2.tidyverse.org/reference/index.html"><code>geom_point()</code></a> layer</p>
<p>Notice that we pipe our tidy data into ggplot2 as before with <code>%\&gt;%</code>. However, for assembling a ggplot script, we instead ‘add’ each component together in series using the <code>+</code>. This happens after initiating the <code>ggplot()</code> function call. After this point, we should for now think of the data as no longer requiring manipulation using function outside of ggplot context. <em>See <a href="https://community.rstudio.com/t/why-cant-ggplot2-use/4372/6">Hadley Wickham’s answer</a> for a better understanding on why ggplot2 uses <code>+</code> and not <code>%\&gt;%</code>.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span> </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="co"># + not %&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> kch_rep1,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">y =</span> kch_rep2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This works, but we notice here that the RNA-seq data is not normally distributed. Instead it is log normal. This means if we want to best visualize a lot of genes at the same time, we should first transform them. We accomplish this by adding 1 to each data point, then taking the log base 2 of each data point: <code>log2(dataExample + 1)</code>. This transform can be easily wrapped into the ggplot code. To learn more about how to use R to compute logarithms and exponents, click this <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/log">link</a> or type <code>?log()</code> into the R console.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log2</span>(kch_rep1 <span class="sc">+</span> <span class="dv">1</span>),</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">y =</span> <span class="fu">log2</span>(kch_rep2 <span class="sc">+</span> <span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Do the same thing but with the other pair of data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log2</span>(kn_rep1 <span class="sc">+</span> <span class="dv">1</span>),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">y =</span> <span class="fu">log2</span>(kn_rep2 <span class="sc">+</span> <span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>But now we want to compare the conditions to each other instead of comparing the biological replicates. To do that we need to summarize the data and compute the mean values. This is accomplished using <code>dplyr::mutate</code>.</p>
<p>First read a little about <a href="https://dplyr.tidyverse.org/reference/mutate.html"><code>mutate()</code></a> and <a href="https://dplyr.tidyverse.org/reference/across.html"><code>across()</code></a>. These are powerful tools for data transforms, but require some practice. They also require careful attention to the logic to ensure operations are occurring as you desire.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">mutate</span>()</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>?<span class="fu">across</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now lets calculate the within group means for our replicates. But before we do that, we should remember we also need out data to be <code>log2(data+1)</code> transformed.The challenge is we need to tell R that we want to transform each cell from the columns we point it to. We can accomplish this by using the vector function <code>dplyr::across</code> inside of mutate. There are many useful strategies for processing data by combining the <code>mutate</code> verb and <code>across</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define the columns to mutate      </span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">.cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># define the function, the '.' represents the data in each cell    </span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">log2</span>(. <span class="sc">+</span> <span class="dv">1</span>) </span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 6
   accession    gene                           kch_rep1 kch_rep2 kn_rep1 kn_rep2
   &lt;chr&gt;        &lt;chr&gt;                             &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2|Ctage5|D12Bwg0579e|Mea6|…     5.48     5.22   4.94    4.80 
 2 NM_172405    Abraxas1|3830405G04Rik|563040…     1.54     1.68   1.56    1.49 
 3 NM_001359283 Sec63|5730478J10Rik|AI649014|…     4.34     4.23   3.83    3.79 
 4 NM_001168290 Sugp2|Sfrs14|Srsf14|mKIAA0365…     3.64     3.30   3.58    3.10 
 5 NM_001356498 Mtmr12|3Pap|4932703C11|C73001…     3.03     3.28   2.95    2.97 
 6 NM_001358950 Cbx5|2610029O15Rik|C75991|HP1…     2.91     2.85   2.78    3.04 
 7 NM_001166375 March1|2900024D24Rik|BB085186…     5.89     6.10   5.66    5.68 
 8 NM_178061    Mob3b|8430436F23Rik|A430018A0…     1.33     1.38   0.790   0.931
 9 NR_126506    Mdm4|4933417N07Rik|AA414968|A…     4.92     4.73   4.12    3.84 
10 NR_029557    Mir145a|Mir145|Mirn145|mir-14…     0        0      0       0    
# … with 40,448 more rows</code></pre>
</div>
</div>
<p>Now pipe the log normalization into a mean calculation, again using <code>dplyr::mutate</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define the columns to mutate      </span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">.cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co"># define the function, the '.' represents the data in each cell      </span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">log2</span>(. <span class="sc">+</span> <span class="dv">1</span>) </span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_mean =</span> (kch_rep1 <span class="sc">+</span> kch_rep2)<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_mean =</span> (kn_rep1 <span class="sc">+</span> kn_rep2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 8
   accession    gene          kch_rep1 kch_rep2 kn_rep1 kn_rep2 kch_mean kn_mean
   &lt;chr&gt;        &lt;chr&gt;            &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2|Ctage5|…     5.48     5.22   4.94    4.80      5.35   4.87 
 2 NM_172405    Abraxas1|383…     1.54     1.68   1.56    1.49      1.61   1.53 
 3 NM_001359283 Sec63|573047…     4.34     4.23   3.83    3.79      4.29   3.81 
 4 NM_001168290 Sugp2|Sfrs14…     3.64     3.30   3.58    3.10      3.47   3.34 
 5 NM_001356498 Mtmr12|3Pap|…     3.03     3.28   2.95    2.97      3.16   2.96 
 6 NM_001358950 Cbx5|2610029…     2.91     2.85   2.78    3.04      2.88   2.91 
 7 NM_001166375 March1|29000…     5.89     6.10   5.66    5.68      6.00   5.67 
 8 NM_178061    Mob3b|843043…     1.33     1.38   0.790   0.931     1.36   0.860
 9 NR_126506    Mdm4|4933417…     4.92     4.73   4.12    3.84      4.82   3.98 
10 NR_029557    Mir145a|Mir1…     0        0      0       0         0      0    
# … with 40,448 more rows</code></pre>
</div>
</div>
<p>Now let’s check our code is working by independently computing a test case. We can identify some test data by ‘slicing’ out a row using the <a href="https://dplyr.tidyverse.org/articles/base.html?q=slice#slice-choose-rows-by-position"><code>dplyr::slice</code></a> verb and selecting the columns to compute.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 2
  kch_rep1 kch_rep2
     &lt;dbl&gt;    &lt;dbl&gt;
1     43.6     36.3</code></pre>
</div>
</div>
<p>Manually calculate the average:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>(<span class="fu">log2</span>(<span class="fl">43.6</span> <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="fu">log2</span>(<span class="fl">36.3</span> <span class="sc">+</span> <span class="dv">1</span>))<span class="sc">/</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.350038</code></pre>
</div>
</div>
<p>Does the value equal what we got above? What is different?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>tb1 <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define the columns to mutate      </span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">.cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co"># define the function, the '.' represents the data in each cell</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">log2</span>(. <span class="sc">+</span> <span class="dv">1</span>) </span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_mean =</span> (kch_rep1 <span class="sc">+</span> kch_rep2)<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_mean =</span> (kn_rep1 <span class="sc">+</span> kn_rep2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  kch_mean
     &lt;dbl&gt;
1     5.35</code></pre>
</div>
</div>
<p>Now lets add these components together and make the comparison scatter plot. Remember that now we are plotting the newly created variable that we generate within the piped commands. Thus we need to provide those variable names to <code>geom_point()</code>. And since we already log transformed, we should not do it again inside the ggplot code block. Note: here I am stashing the transformed data as a new variable. I then use this new variable to pipe into ggplot on a new coding line. We will use this variable further below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>tb2 <span class="ot">&lt;-</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  tb1 <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">.cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">log2</span>(. <span class="sc">+</span> <span class="dv">1</span>) </span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_mean =</span> (kch_rep1 <span class="sc">+</span> kch_rep2)<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_mean =</span> (kn_rep1 <span class="sc">+</span> kn_rep2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span></span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Ok great, but now we really want to know the identity of the genes with the largest differences! Let’s make a new variable with only those genes. We normally accomplish this using differential expression analysis. Today let’s use a simpler strategy by just comparing the fold-change.</p>
<p>Remember, in <a href="https://en.wikipedia.org/wiki/Logarithm">logarithmic mathematics</a>, <code>log2(x+1) - log2(y+1)</code> is equal to <code>log2((x+1) / (y+1))</code>. Since our data is already log transformed, the fold change is computer by comparing the difference of the numbers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">logFC =</span> kch_mean <span class="sc">-</span> kn_mean</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 40,458 × 9
   accession    gene  kch_rep1 kch_rep2 kn_rep1 kn_rep2 kch_mean kn_mean   logFC
   &lt;chr&gt;        &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2…     5.48     5.22   4.94    4.80      5.35   4.87   0.483 
 2 NM_172405    Abra…     1.54     1.68   1.56    1.49      1.61   1.53   0.0830
 3 NM_001359283 Sec6…     4.34     4.23   3.83    3.79      4.29   3.81   0.474 
 4 NM_001168290 Sugp…     3.64     3.30   3.58    3.10      3.47   3.34   0.136 
 5 NM_001356498 Mtmr…     3.03     3.28   2.95    2.97      3.16   2.96   0.196 
 6 NM_001358950 Cbx5…     2.91     2.85   2.78    3.04      2.88   2.91  -0.0253
 7 NM_001166375 Marc…     5.89     6.10   5.66    5.68      6.00   5.67   0.330 
 8 NM_178061    Mob3…     1.33     1.38   0.790   0.931     1.36   0.860  0.496 
 9 NR_126506    Mdm4…     4.92     4.73   4.12    3.84      4.82   3.98   0.845 
10 NR_029557    Mir1…     0        0      0       0         0      0      0     
# … with 40,448 more rows</code></pre>
</div>
</div>
<p>Ok, now lets identify the instance were the logFC is &gt; 1 or &lt; -1. In other words, a fold change greater than ‘2’. For this we will use a new dplyr verb, <a href="https://dplyr.tidyverse.org/reference/filter.html"><code>dplyr::filter</code></a>, which operates on data rows. We also will need to use an R logical operator. With R, <code>|</code> is equivalent to or, <code>&amp;</code> is equivalent to and. Read more here: <a href="https://www.statmethods.net/management/operators.html" class="uri">https://www.statmethods.net/management/operators.html</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">logFC =</span> kch_mean <span class="sc">-</span> kn_mean</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,757 × 9
   accession    gene    kch_rep1 kch_rep2 kn_rep1 kn_rep2 kch_mean kn_mean logFC
   &lt;chr&gt;        &lt;chr&gt;      &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
 1 NM_021334    Itgax|…    2.06     2.05     6.58    7.04    2.06     6.81 -4.75
 2 NM_001111059 Cd34|A…    0.579    0.573    2.06    1.86    0.576    1.96 -1.39
 3 NM_022995    Pmepa1…    0.864    1.06     1.89    2.61    0.961    2.25 -1.29
 4 NM_008433    Kcnn4|…    0.840    1.08     3.38    3.75    0.959    3.56 -2.60
 5 NM_001130458 Tcn2|A…    9.61     9.87     8.33    7.84    9.74     8.09  1.65
 6 NR_046233    Rn45s|…   10.9     11.1      8.59   10.5    11.0      9.56  1.46
 7 NM_010730    Anxa1|…    0.766    1.05     3.25    3.90    0.906    3.57 -2.67
 8 NR_122039    Tsc22d…    6.68     6.90     6.09    5.27    6.79     5.68  1.11
 9 NM_001361978 Arsg|6…    2.12     2.56     1.50    1.07    2.34     1.29  1.05
10 NM_145134    Spsb4|…    0.327    0.842    3.03    3.39    0.584    3.21 -2.63
# … with 1,747 more rows</code></pre>
</div>
</div>
<p>This yields a lot of ‘different’ genes! Lets make our space smaller by being more restrictive. Lets also generate a new variable with this smaller data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>tb3 <span class="ot">&lt;-</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">logFC =</span> kch_mean <span class="sc">-</span> kn_mean</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">6</span>)</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>tb3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 9
  accession    gene     kch_rep1 kch_rep2 kn_rep1 kn_rep2 kch_mean kn_mean logFC
  &lt;chr&gt;        &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 NM_053110    Gpnmb|D…    1.03     0.969    8.08    8.79    1.00     8.44 -7.44
2 NM_018866    Cxcl13|…   10.3      7.21     3.80    1.46    8.75     2.63  6.12
3 NM_008605    Mmp12|A…    1.88     2.18     8.02    8.10    2.03     8.06 -6.03
4 NM_053094    Cd163|C…   10.6     10.7      4.85    3.10   10.6      3.97  6.67
5 NM_001320076 Mmp12|A…    1.87     2.17     8.02    8.09    2.02     8.05 -6.04
6 NM_001204203 Spp1|2A…    0.597    0.570    5.79    7.39    0.584    6.59 -6.00
7 NM_009263    Spp1|2A…    0.598    0.571    5.78    7.39    0.585    6.59 -6.00
8 NM_001170395 Cd163|C…   10.5     10.7      4.83    3.08   10.6      3.95  6.66</code></pre>
</div>
</div>
<p>Ok, 8 genes is a small enough set for now. Now lets add labels to our scatter plot above with the identifier information for these genes. We can accomplish this in several ways. One way is to overlay the text from the new filtered variable using <a href="https://ggplot2.tidyverse.org/articles/articles/faq-annotation.html?q=geom%20_%20text#why-is-annotation-created-with-geom_text-pixellated-how-can-i-make-it-more-crisp"><code>geom_text()</code></a>. In this case we must define the source of the data and the x and y positions must be the same. We also must supply the identity of the column containing the labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean)) <span class="sc">+</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> tb3, <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> accession))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Ok, that’s not bad, but a gene name would be better. Unfortunately, our gene names are buried in with a lot of information. Let’s extract it! That is made easy with columns using the <a href="https://tidyr.tidyverse.org/reference/separate.html"><code>tidyr::separate</code></a> verb. Read about it first:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>?tidyr<span class="sc">::</span><span class="fu">separate</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>tb3 <span class="ot">&lt;-</span> </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">logFC =</span> kch_mean <span class="sc">-</span> kn_mean</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(<span class="at">col =</span> <span class="dv">2</span>, <span class="at">into =</span> <span class="st">"gene"</span>, </span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="co"># this part is tricky and follow 'regular expression' rules           </span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">sep =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>, </span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>           <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">extra =</span> <span class="st">"drop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now lets regenerate the scatter plot with the gene names labeled instead.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span> </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean)) <span class="sc">+</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> tb3, <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> gene))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Pretty good, but those duplicate gene names from splice variant accessions are cause the plot to look blurry. We can remove those without much forethought using the <a href="https://dplyr.tidyverse.org/reference/distinct.html"><code>dplyr::distinct</code></a> verb. That’s not the best idea long term, but useful for now. I will apply this to the initial data being plotted and also inside <code>geom_text()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">distinct</span>()</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>tb2 <span class="sc">%&gt;%</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean)) <span class="sc">+</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> tb3 <span class="sc">%&gt;%</span> <span class="fu">distinct</span>(gene, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>),</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> gene))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now we can stack it all together. Into one tibble. And use that tibble for our plot. Which also provides the opportunity to illustrate that dplyr functions can be layered inside of gglot functions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>tb4 <span class="ot">&lt;-</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  readxl<span class="sc">::</span><span class="fu">read_xlsx</span>(<span class="at">path =</span> <span class="st">"../Data/Source/1-s2.0-S107476132030159X-mmc3.xlsx"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>) <span class="sc">%&gt;%</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">accession =</span> <span class="dv">1</span>,</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene =</span> <span class="dv">2</span>,</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_rep1 =</span> <span class="dv">3</span>,</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_rep2 =</span> <span class="dv">4</span>,</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_rep1 =</span> <span class="dv">5</span>,</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_rep2 =</span> <span class="dv">6</span></span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(<span class="at">col =</span> <span class="dv">2</span>, <span class="at">into =</span> <span class="st">"gene"</span>, </span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a><span class="co"># this part is tricky and follow 'regular expression' rules</span></span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">sep =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>, </span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>           <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">extra =</span> <span class="st">"drop"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb58-17"><a href="#cb58-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb58-18"><a href="#cb58-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(</span>
<span id="cb58-19"><a href="#cb58-19" aria-hidden="true" tabindex="-1"></a><span class="co"># define the columns to mutate      </span></span>
<span id="cb58-20"><a href="#cb58-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">.cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>, </span>
<span id="cb58-21"><a href="#cb58-21" aria-hidden="true" tabindex="-1"></a><span class="co"># define the function, the '.' represents the data in each cell      </span></span>
<span id="cb58-22"><a href="#cb58-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">.fns =</span> <span class="sc">~</span> <span class="fu">log2</span>(. <span class="sc">+</span> <span class="dv">1</span>) </span>
<span id="cb58-23"><a href="#cb58-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb58-24"><a href="#cb58-24" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb58-25"><a href="#cb58-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb58-26"><a href="#cb58-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">kch_mean =</span> (kch_rep1 <span class="sc">+</span> kch_rep2)<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb58-27"><a href="#cb58-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">kn_mean =</span> (kn_rep1 <span class="sc">+</span> kn_rep2)<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb58-28"><a href="#cb58-28" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb58-29"><a href="#cb58-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb58-30"><a href="#cb58-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">logFC =</span> kch_mean <span class="sc">-</span> kn_mean</span>
<span id="cb58-31"><a href="#cb58-31" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb58-32"><a href="#cb58-32" aria-hidden="true" tabindex="-1"></a>tb4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 24,941 × 9
   accession    gene  kch_rep1 kch_rep2 kn_rep1 kn_rep2 kch_mean kn_mean   logFC
   &lt;chr&gt;        &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 NM_001329047 Mia2      5.48     5.22   4.94    4.80      5.35   4.87   0.483 
 2 NM_172405    Abra…     1.54     1.68   1.56    1.49      1.61   1.53   0.0830
 3 NM_001359283 Sec63     4.34     4.23   3.83    3.79      4.29   3.81   0.474 
 4 NM_001168290 Sugp2     3.64     3.30   3.58    3.10      3.47   3.34   0.136 
 5 NM_001356498 Mtmr…     3.03     3.28   2.95    2.97      3.16   2.96   0.196 
 6 NM_001358950 Cbx5      2.91     2.85   2.78    3.04      2.88   2.91  -0.0253
 7 NM_001166375 Marc…     5.89     6.10   5.66    5.68      6.00   5.67   0.330 
 8 NM_178061    Mob3b     1.33     1.38   0.790   0.931     1.36   0.860  0.496 
 9 NR_126506    Mdm4      4.92     4.73   4.12    3.84      4.82   3.98   0.845 
10 NR_029557    Mir1…     0        0      0       0         0      0      0     
# … with 24,931 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>tb4 <span class="sc">%&gt;%</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean)) <span class="sc">+</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> tb4 <span class="sc">%&gt;%</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>              <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">6</span>), </span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> gene))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally, we have a plot that we may want to share. This is easy with ggplot using <a href="https://ggplot2.tidyverse.org/reference/ggsave.html?q=ggsave"><code>ggave()</code></a>. We can either make the plot itself into a variable, or simply generate an image using the last ggplot created.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  tb4 <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean)) <span class="sc">+</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> tb4 <span class="sc">%&gt;%</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>              <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">6</span>),</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> gene))</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"../Data/Plots/scatterchart1.png"</span>,</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> plot1,</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">4</span>,</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">height =</span> <span class="dv">4</span>,</span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">units =</span> <span class="st">"in"</span>,</span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">dpi =</span> <span class="dv">150</span>,</span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg =</span> <span class="st">"white"</span></span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The plot can be refined in many ways. And the text can be better controlled using the <a href="https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html">ggrepel package</a>. See what you can do!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ggrepel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggrepel</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>tb4 <span class="sc">%&gt;%</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(kch_mean <span class="sc">&gt;</span> <span class="dv">2</span> <span class="sc">|</span> kn_mean <span class="sc">&gt;</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bin_2d</span>(<span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean), <span class="at">bins =</span> <span class="dv">101</span>) <span class="sc">+</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_continuous</span>(<span class="at">type =</span> <span class="st">"viridis"</span>) <span class="sc">+</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> tb4 <span class="sc">%&gt;%</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(logFC <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">|</span> logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">6</span>),</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> kch_mean, <span class="at">y =</span> kn_mean, <span class="at">label =</span> gene),</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">fontface =</span> <span class="st">"italic"</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, </span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Healthy Kupffer cells"</span>, </span>
<span id="cb64-16"><a href="#cb64-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"NASH Kupffer cells"</span>, </span>
<span id="cb64-17"><a href="#cb64-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"RNA-seq, mean expression"</span>) <span class="sc">+</span></span>
<span id="cb64-18"><a href="#cb64-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Introduction-RNASeqAnalysis_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"../Data/Plots/scatterchart2.png"</span>,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">4</span>,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">height =</span> <span class="dv">4</span>,</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">units =</span> <span class="st">"in"</span>,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">dpi =</span> <span class="dv">150</span>,</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg =</span> <span class="st">"white"</span></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="finished" class="level1">
<h1>Finished!!</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">pkgs =</span> <span class="cn">NULL</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  details<span class="sc">::</span><span class="fu">details</span>(<span class="at">summary =</span> <span class="st">'Current session info'</span>, <span class="at">open =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<details open="">
<summary>
<span title="Click to Expand"> Current session info </span>
</summary>
<div class="sourceCode" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>─ Session info ───────────────────────────────────────────────────────────────</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a> setting  value</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a> version  R version <span class="dv">4</span>.<span class="fl">1.0</span> (<span class="dv">2021-05-18</span>)</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a> os       macOS Big Sur <span class="fl">10.16</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a> system   x86_64, darwin17<span class="fl">.0</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a> ui       X11</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a> <span class="fu">language</span> (EN)</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a> collate  en_US.UTF<span class="dv">-8</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a> ctype    en_US.UTF<span class="dv">-8</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a> tz       America<span class="sc">/</span>New_York</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a> date     <span class="dv">2022-05-05</span></span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a> pandoc   <span class="dv">2</span>.<span class="dv">17</span>.<span class="fl">1.1</span> <span class="sc">@</span> <span class="er">/</span>Applications<span class="sc">/</span>RStudio.app<span class="sc">/</span>Contents<span class="sc">/</span>MacOS<span class="sc">/</span>quarto<span class="sc">/</span>bin<span class="sc">/</span> (via rmarkdown)</span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a>─ Packages ───────────────────────────────────────────────────────────────────</span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a> package     <span class="sc">*</span> version <span class="fu">date</span> (UTC) lib source</span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a> assertthat    <span class="dv">0</span>.<span class="fl">2.1</span>   <span class="dv">2019-03-21</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-18"><a href="#cb67-18" aria-hidden="true" tabindex="-1"></a> backports     <span class="dv">1</span>.<span class="fl">4.1</span>   <span class="dv">2021-12-13</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-19"><a href="#cb67-19" aria-hidden="true" tabindex="-1"></a> broom         <span class="dv">0</span>.<span class="fl">7.12</span>  <span class="dv">2022-01-28</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-20"><a href="#cb67-20" aria-hidden="true" tabindex="-1"></a> cellranger    <span class="dv">1</span>.<span class="fl">1.0</span>   <span class="dv">2016-07-27</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-21"><a href="#cb67-21" aria-hidden="true" tabindex="-1"></a> cli           <span class="dv">3</span>.<span class="fl">1.1</span>   <span class="dv">2022-01-20</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-22"><a href="#cb67-22" aria-hidden="true" tabindex="-1"></a> clipr         <span class="dv">0</span>.<span class="fl">7.1</span>   <span class="dv">2020-10-08</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-23"><a href="#cb67-23" aria-hidden="true" tabindex="-1"></a> colorspace    <span class="fl">2.0</span><span class="dv">-2</span>   <span class="dv">2021-06-24</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-24"><a href="#cb67-24" aria-hidden="true" tabindex="-1"></a> crayon        <span class="dv">1</span>.<span class="fl">4.2</span>   <span class="dv">2021-10-29</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-25"><a href="#cb67-25" aria-hidden="true" tabindex="-1"></a> DBI           <span class="dv">1</span>.<span class="fl">1.2</span>   <span class="dv">2021-12-20</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-26"><a href="#cb67-26" aria-hidden="true" tabindex="-1"></a> dbplyr        <span class="dv">2</span>.<span class="fl">1.1</span>   <span class="dv">2021-04-06</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-27"><a href="#cb67-27" aria-hidden="true" tabindex="-1"></a> desc          <span class="dv">1</span>.<span class="fl">4.0</span>   <span class="dv">2021-09-28</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-28"><a href="#cb67-28" aria-hidden="true" tabindex="-1"></a> details       <span class="dv">0</span>.<span class="fl">2.1</span>   <span class="dv">2020-01-12</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-29"><a href="#cb67-29" aria-hidden="true" tabindex="-1"></a> digest        <span class="dv">0</span>.<span class="fl">6.29</span>  <span class="dv">2021-12-01</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-30"><a href="#cb67-30" aria-hidden="true" tabindex="-1"></a> dplyr       <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">0.8</span>   <span class="dv">2022-02-08</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-31"><a href="#cb67-31" aria-hidden="true" tabindex="-1"></a> ellipsis      <span class="dv">0</span>.<span class="fl">3.2</span>   <span class="dv">2021-04-29</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-32"><a href="#cb67-32" aria-hidden="true" tabindex="-1"></a> evaluate      <span class="fl">0.14</span>    <span class="dv">2019-05-28</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-33"><a href="#cb67-33" aria-hidden="true" tabindex="-1"></a> fansi         <span class="dv">1</span>.<span class="fl">0.2</span>   <span class="dv">2022-01-14</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-34"><a href="#cb67-34" aria-hidden="true" tabindex="-1"></a> farver        <span class="dv">2</span>.<span class="fl">1.0</span>   <span class="dv">2021-02-28</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-35"><a href="#cb67-35" aria-hidden="true" tabindex="-1"></a> fastmap       <span class="dv">1</span>.<span class="fl">1.0</span>   <span class="dv">2021-01-25</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-36"><a href="#cb67-36" aria-hidden="true" tabindex="-1"></a> forcats     <span class="sc">*</span> <span class="dv">0</span>.<span class="fl">5.1</span>   <span class="dv">2021-01-27</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-37"><a href="#cb67-37" aria-hidden="true" tabindex="-1"></a> fs            <span class="dv">1</span>.<span class="fl">5.2</span>   <span class="dv">2021-12-08</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-38"><a href="#cb67-38" aria-hidden="true" tabindex="-1"></a> generics      <span class="dv">0</span>.<span class="fl">1.2</span>   <span class="dv">2022-01-31</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-39"><a href="#cb67-39" aria-hidden="true" tabindex="-1"></a> ggplot2     <span class="sc">*</span> <span class="dv">3</span>.<span class="fl">3.5</span>   <span class="dv">2021-06-25</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-40"><a href="#cb67-40" aria-hidden="true" tabindex="-1"></a> ggrepel     <span class="sc">*</span> <span class="dv">0</span>.<span class="fl">9.1</span>   <span class="dv">2021-01-15</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-41"><a href="#cb67-41" aria-hidden="true" tabindex="-1"></a> glue          <span class="dv">1</span>.<span class="fl">6.1</span>   <span class="dv">2022-01-22</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-42"><a href="#cb67-42" aria-hidden="true" tabindex="-1"></a> gtable        <span class="dv">0</span>.<span class="fl">3.0</span>   <span class="dv">2019-03-25</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-43"><a href="#cb67-43" aria-hidden="true" tabindex="-1"></a> haven         <span class="dv">2</span>.<span class="fl">4.3</span>   <span class="dv">2021-08-04</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-44"><a href="#cb67-44" aria-hidden="true" tabindex="-1"></a> hms           <span class="dv">1</span>.<span class="fl">1.1</span>   <span class="dv">2021-09-26</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-45"><a href="#cb67-45" aria-hidden="true" tabindex="-1"></a> htmltools     <span class="dv">0</span>.<span class="fl">5.2</span>   <span class="dv">2021-08-25</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-46"><a href="#cb67-46" aria-hidden="true" tabindex="-1"></a> htmlwidgets   <span class="dv">1</span>.<span class="fl">5.4</span>   <span class="dv">2021-09-08</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-47"><a href="#cb67-47" aria-hidden="true" tabindex="-1"></a> httr          <span class="dv">1</span>.<span class="fl">4.2</span>   <span class="dv">2020-07-20</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-48"><a href="#cb67-48" aria-hidden="true" tabindex="-1"></a> jsonlite      <span class="dv">1</span>.<span class="fl">7.3</span>   <span class="dv">2022-01-17</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-49"><a href="#cb67-49" aria-hidden="true" tabindex="-1"></a> knitr         <span class="fl">1.37</span>    <span class="dv">2021-12-16</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-50"><a href="#cb67-50" aria-hidden="true" tabindex="-1"></a> labeling      <span class="dv">0</span>.<span class="fl">4.2</span>   <span class="dv">2020-10-20</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-51"><a href="#cb67-51" aria-hidden="true" tabindex="-1"></a> lifecycle     <span class="dv">1</span>.<span class="fl">0.1</span>   <span class="dv">2021-09-24</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-52"><a href="#cb67-52" aria-hidden="true" tabindex="-1"></a> lubridate     <span class="dv">1</span>.<span class="fl">8.0</span>   <span class="dv">2021-10-07</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-53"><a href="#cb67-53" aria-hidden="true" tabindex="-1"></a> magrittr      <span class="dv">2</span>.<span class="fl">0.2</span>   <span class="dv">2022-01-26</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-54"><a href="#cb67-54" aria-hidden="true" tabindex="-1"></a> modelr        <span class="dv">0</span>.<span class="fl">1.8</span>   <span class="dv">2020-05-19</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-55"><a href="#cb67-55" aria-hidden="true" tabindex="-1"></a> munsell       <span class="dv">0</span>.<span class="fl">5.0</span>   <span class="dv">2018-06-12</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-56"><a href="#cb67-56" aria-hidden="true" tabindex="-1"></a> pillar        <span class="dv">1</span>.<span class="fl">7.0</span>   <span class="dv">2022-02-01</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-57"><a href="#cb67-57" aria-hidden="true" tabindex="-1"></a> pkgconfig     <span class="dv">2</span>.<span class="fl">0.3</span>   <span class="dv">2019-09-22</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-58"><a href="#cb67-58" aria-hidden="true" tabindex="-1"></a> png           <span class="fl">0.1</span><span class="dv">-7</span>   <span class="dv">2013-12-03</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-59"><a href="#cb67-59" aria-hidden="true" tabindex="-1"></a> purrr       <span class="sc">*</span> <span class="dv">0</span>.<span class="fl">3.4</span>   <span class="dv">2020-04-17</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-60"><a href="#cb67-60" aria-hidden="true" tabindex="-1"></a> R6            <span class="dv">2</span>.<span class="fl">5.1</span>   <span class="dv">2021-08-19</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-61"><a href="#cb67-61" aria-hidden="true" tabindex="-1"></a> ragg          <span class="dv">1</span>.<span class="fl">1.3</span>   <span class="dv">2021-06-09</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-62"><a href="#cb67-62" aria-hidden="true" tabindex="-1"></a> Rcpp          <span class="dv">1</span>.<span class="fl">0.8</span>   <span class="dv">2022-01-13</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-63"><a href="#cb67-63" aria-hidden="true" tabindex="-1"></a> readr       <span class="sc">*</span> <span class="dv">2</span>.<span class="fl">1.2</span>   <span class="dv">2022-01-30</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-64"><a href="#cb67-64" aria-hidden="true" tabindex="-1"></a> readxl        <span class="dv">1</span>.<span class="fl">3.1</span>   <span class="dv">2019-03-13</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-65"><a href="#cb67-65" aria-hidden="true" tabindex="-1"></a> reprex        <span class="dv">2</span>.<span class="fl">0.1</span>   <span class="dv">2021-08-05</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-66"><a href="#cb67-66" aria-hidden="true" tabindex="-1"></a> rlang         <span class="dv">1</span>.<span class="fl">0.1</span>   <span class="dv">2022-02-03</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-67"><a href="#cb67-67" aria-hidden="true" tabindex="-1"></a> rmarkdown     <span class="fl">2.11</span>    <span class="dv">2021-09-14</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-68"><a href="#cb67-68" aria-hidden="true" tabindex="-1"></a> rprojroot     <span class="dv">2</span>.<span class="fl">0.2</span>   <span class="dv">2020-11-15</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-69"><a href="#cb67-69" aria-hidden="true" tabindex="-1"></a> rstudioapi    <span class="fl">0.13</span>    <span class="dv">2020-11-12</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-70"><a href="#cb67-70" aria-hidden="true" tabindex="-1"></a> rvest         <span class="dv">1</span>.<span class="fl">0.2</span>   <span class="dv">2021-10-16</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-71"><a href="#cb67-71" aria-hidden="true" tabindex="-1"></a> scales        <span class="dv">1</span>.<span class="fl">1.1</span>   <span class="dv">2020-05-11</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-72"><a href="#cb67-72" aria-hidden="true" tabindex="-1"></a> sessioninfo   <span class="dv">1</span>.<span class="fl">2.2</span>   <span class="dv">2021-12-06</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-73"><a href="#cb67-73" aria-hidden="true" tabindex="-1"></a> stringi       <span class="dv">1</span>.<span class="fl">7.6</span>   <span class="dv">2021-11-29</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-74"><a href="#cb67-74" aria-hidden="true" tabindex="-1"></a> stringr     <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">4.0</span>   <span class="dv">2019-02-10</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-75"><a href="#cb67-75" aria-hidden="true" tabindex="-1"></a> systemfonts   <span class="dv">1</span>.<span class="fl">0.2</span>   <span class="dv">2021-05-11</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-76"><a href="#cb67-76" aria-hidden="true" tabindex="-1"></a> textshaping   <span class="dv">0</span>.<span class="fl">3.6</span>   <span class="dv">2021-10-13</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-77"><a href="#cb67-77" aria-hidden="true" tabindex="-1"></a> tibble      <span class="sc">*</span> <span class="dv">3</span>.<span class="fl">1.6</span>   <span class="dv">2021-11-07</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-78"><a href="#cb67-78" aria-hidden="true" tabindex="-1"></a> tidyr       <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">2.0</span>   <span class="dv">2022-02-01</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-79"><a href="#cb67-79" aria-hidden="true" tabindex="-1"></a> tidyselect    <span class="dv">1</span>.<span class="fl">1.1</span>   <span class="dv">2021-04-30</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-80"><a href="#cb67-80" aria-hidden="true" tabindex="-1"></a> tidyverse   <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">3.1</span>   <span class="dv">2021-04-15</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-81"><a href="#cb67-81" aria-hidden="true" tabindex="-1"></a> tzdb          <span class="dv">0</span>.<span class="fl">2.0</span>   <span class="dv">2021-10-27</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-82"><a href="#cb67-82" aria-hidden="true" tabindex="-1"></a> utf8          <span class="dv">1</span>.<span class="fl">2.2</span>   <span class="dv">2021-07-24</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-83"><a href="#cb67-83" aria-hidden="true" tabindex="-1"></a> vctrs         <span class="dv">0</span>.<span class="fl">3.8</span>   <span class="dv">2021-04-29</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-84"><a href="#cb67-84" aria-hidden="true" tabindex="-1"></a> viridisLite   <span class="dv">0</span>.<span class="fl">4.0</span>   <span class="dv">2021-04-13</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-85"><a href="#cb67-85" aria-hidden="true" tabindex="-1"></a> withr         <span class="dv">2</span>.<span class="fl">4.3</span>   <span class="dv">2021-11-30</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-86"><a href="#cb67-86" aria-hidden="true" tabindex="-1"></a> xfun          <span class="fl">0.29</span>    <span class="dv">2021-12-14</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-87"><a href="#cb67-87" aria-hidden="true" tabindex="-1"></a> xml2          <span class="dv">1</span>.<span class="fl">3.3</span>   <span class="dv">2021-11-30</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.0</span>)</span>
<span id="cb67-88"><a href="#cb67-88" aria-hidden="true" tabindex="-1"></a> yaml          <span class="dv">2</span>.<span class="fl">2.2</span>   <span class="dv">2022-01-25</span> [<span class="dv">2</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">1.2</span>)</span>
<span id="cb67-89"><a href="#cb67-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-90"><a href="#cb67-90" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] <span class="sc">/</span>Users<span class="sc">/</span>tro3nr<span class="sc">/</span>Library<span class="sc">/</span>R<span class="sc">/</span>x86_64<span class="sc">/</span><span class="fl">4.1</span><span class="sc">/</span>library</span>
<span id="cb67-91"><a href="#cb67-91" aria-hidden="true" tabindex="-1"></a> [<span class="dv">2</span>] <span class="sc">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.1</span><span class="sc">/</span>Resources<span class="sc">/</span>library</span>
<span id="cb67-92"><a href="#cb67-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-93"><a href="#cb67-93" aria-hidden="true" tabindex="-1"></a>──────────────────────────────────────────────────────────────────────────────</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<p><br></p>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
    } else {
      disableStylesheet(alternateStylesheets);
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } 
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>