<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Troutman Lab - RNA-seq analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Troutman Lab</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../qmd/Internal_Materials.html"> 
<span class="menu-text">Internal Materials</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ty-troutman"> <i class="bi bi-github" role="img" aria-label="Troutman GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://www.cincinnatichildrens.org/research/divisions/a/allergy-immunology/labs/troutman"> 
<span class="menu-text">CCHMC</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#general-workflow" id="toc-general-workflow" class="nav-link active" data-scroll-target="#general-workflow">General workflow</a>
  <ul class="collapse">
  <li><a href="#data-processing" id="toc-data-processing" class="nav-link" data-scroll-target="#data-processing">Data processing</a></li>
  <li><a href="#data-quality-control" id="toc-data-quality-control" class="nav-link" data-scroll-target="#data-quality-control">Data quality control</a></li>
  <li><a href="#data-quantification" id="toc-data-quantification" class="nav-link" data-scroll-target="#data-quantification">Data quantification</a></li>
  <li><a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis">Differential expression analysis</a></li>
  <li><a href="#analytical-creativity" id="toc-analytical-creativity" class="nav-link" data-scroll-target="#analytical-creativity">Analytical creativity</a></li>
  </ul></li>
  <li><a href="#automated-pipelines" id="toc-automated-pipelines" class="nav-link" data-scroll-target="#automated-pipelines">Automated pipelines</a>
  <ul class="collapse">
  <li><a href="#map_on_bmi_hpc-in-progress" id="toc-map_on_bmi_hpc-in-progress" class="nav-link" data-scroll-target="#map_on_bmi_hpc-in-progress">map_on_bmi_hpc (in progress!)</a></li>
  <li><a href="#a.c.rsuite" id="toc-a.c.rsuite" class="nav-link" data-scroll-target="#a.c.rsuite">A.C.Rsuite</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">RNA-seq analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This is meant to be a guideline for processing and analysis of RNA-seq data in the Troutman lab. There are many alternative ways to consider processing, quality control, quantification, and differential expression analysis. This workflow is not a gold standard, will not always work for all data, and is open for interpretation/criticism.</p>
<section id="general-workflow" class="level2">
<h2 class="anchored" data-anchor-id="general-workflow">General workflow</h2>
<section id="data-processing" class="level3">
<h3 class="anchored" data-anchor-id="data-processing">Data processing</h3>
<ol type="1">
<li><p>Verify data checksums and rename following lab convention: &lt;filenameA&gt;_R1.fastq.gz &lt;filenameA&gt;_R2.fastq.gz &lt;filenameB&gt;_R1.fastq.gz &lt;filenameB&gt;_R2.fastq.gz</p></li>
<li><p>Group data according to mapping needs, e.g.&nbsp;by genome, paired end and single read, etc.</p></li>
<li><p>Run <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a> on the raw fastq files.</p></li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc/0.11.7</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="op">&lt;</span>data<span class="op">&gt;</span>.R1.fastq.gz <span class="at">-o</span> <span class="op">&lt;</span>preferred_name<span class="op">&gt;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="op">&lt;</span>data<span class="op">&gt;</span>.R2.fastq.gz <span class="at">-o</span> <span class="op">&lt;</span>preferred_name<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="4" type="1">
<li>Run <a href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim Galore</a> on the fastq files. Rerun fastqc on the trimmed fastq files. Trim Galore can automatically do this for you.</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load trimgalore/0.6.6</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--cores</span> 4 <span class="dt">\</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--paired</span> <span class="op">&lt;</span>data_R1.fastq.gz<span class="op">&gt;</span> <span class="op">&lt;</span>data_R2.fastq.gz<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastqc</span> <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output_dir</span> <span class="op">&lt;</span>preferred_path<span class="op">&gt;</span>/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="5" type="1">
<li>Align the data with <a href="https://github.com/alexdobin/STAR">STAR</a>. Have STAR format the alignment as a sorted BAM file.</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load STAR/2.7.9</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load samtools</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--runThreadN</span> 16 <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genomeDir</span> /path/to/star/index/ <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesIn</span> <span class="op">&lt;</span>data_R1.fastq.gz<span class="op">&gt;</span> <span class="op">&lt;</span>data_R2.fastq.gz<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> <span class="op">&lt;</span>preferred_prefix<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="6" type="1">
<li>Use <a href="http://homer.ucsd.edu/homer/">HOMER</a> to generate <a href="http://homer.ucsd.edu/homer/ngs/tagDir.html">tag directories</a> from the STAR alignment BAM file. Use the <code>-sspe -flip</code> options if the data is generated using our lab’s dUTP protocol, which is strand specific and reports on the opposite strand from the original molecule.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">makeTagDirectory</span> <span class="op">&lt;</span>tagDirectoryName<span class="op">&gt;</span> <span class="op">&lt;</span>filename.sam<span class="op">&gt;</span> -genome <span class="op">&lt;</span>experiment_genome<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-checkGC</span> <span class="at">-format</span> sam <span class="at">-sspe</span> <span class="at">-flip</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Omit `-sspe` if the sequencing data is not paired end, and omit `-flip` if the sequencing reads will not align to the opposite DNA strand as in "first-strand" library synthesis protocols.</code></pre>
<ol start="7" type="1">
<li>Use <a href="http://www.htslib.org">Samtools</a> to <a href="http://www.htslib.org/doc/samtools-index.html">index</a> the sorted BAM files.</li>
</ol>
</section>
<section id="data-quality-control" class="level3">
<h3 class="anchored" data-anchor-id="data-quality-control">Data quality control</h3>
<ol type="1">
<li>Use the fastqc results to check basic data quality.</li>
<li>Check data assumptions and quality using a local install of the <a href="https://software.broadinstitute.org/software/igv/">Integrative Genomics Viewer</a>. An alternative would be to load a HOMER <a href="http://homer.ucsd.edu/homer/ngs/ucsc.html"><code>makeUCSCfile</code></a> into the <a href="https://genome.ucsc.edu">UCSC Genome Browser</a>. An even better solution is to load HOMER <a href="http://homer.ucsd.edu/homer/ngs/ucsc.html"><code>makeMultiWigHub.pl</code></a> tracks to the UCSC Genome Browser using a lab web server. <em>We have a web server for this purpose, but the resource is not functional yet.</em></li>
<li>Check mapping efficiency using the STAR alignment log file.</li>
<li>Check other basic quality parameters using the HOMER <code>tagInfo.txt</code>, <code>tagLengthDistribution.txt</code>, and <code>tagCountDistribution.txt</code> files. These files will be found in the HOMER tag directory.</li>
<li>Check sequence bias using the HOMER <code>tagFreq.txt</code> and <code>tagFreqUniq.txt</code>. These files will also be in the HOMER tag directory. Fastqc also reports on sequencing bias.</li>
</ol>
<p>Many of these QC parameters can quickly be summarized using <a href="https://multiqc.info">MultiQC</a>. This software is not currently available on the BMI cluster but can be run locally. Installation is easy via <a href="https://www.docker.com">Docker</a>.</p>
<p>In order to run MultiQc from your local machine, you will need to do one of the following: mount the server folder as a drive on your local machine (good idea), or copy the data from the server to your local machine (bad idea). Instructions for mounting the drives can be found <a href="../qmd/server_lab_drives.html">here</a> and <a href="https://confluence.research.cchmc.org/display/RESITHUB/Data+Storage">here</a>.</p>
<p>Once a drive is mounted and the Docker software is opened, the Multiqc docker image can be run through your terminal application on your local machine as in the example described <a href="https://multiqc.info/docs/getting_started/installation/#docker">here</a> and provided below:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-t</span> <span class="at">-v</span> <span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span>:<span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span> <span class="at">-w</span> <span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span> ewels/multiqc multiqc ./path/to/mounted/drive</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Another useful tool is <code>alignStats.R</code>, which generates a table containing selected information from the STAR log file and the HOMER <code>tagInfo.txt</code> file. This tool is described in more detail <a href="map_on_bmi_hpc (in progress!)">below</a>.</p>
</section>
<section id="data-quantification" class="level3">
<h3 class="anchored" data-anchor-id="data-quantification">Data quantification</h3>
<ol type="1">
<li>Quantify sequencing reads using HOMER <code>[analyzeRepeat.pl]</code>(http://homer.ucsd.edu/homer/ngs/analyzeRNA.html).</li>
<li>Compare gene detection depth by plotting distribution of normalized transcript counts at various thresholds.</li>
<li>Validate experimental replicates and treatment differences with log transformed data using principal component analysis and/or pairwise correlation analysis.</li>
<li>Compare expression of some control and test genes using bar plots, box whisker plots, and/or a genome browser.</li>
<li>Compare global expression profiles with log transformed data by heat map. Assess global expression profiles with and without row scaling.</li>
</ol>
</section>
<section id="differential-expression-analysis" class="level3">
<h3 class="anchored" data-anchor-id="differential-expression-analysis">Differential expression analysis</h3>
<ol type="1">
<li>Differential expression analysis should be used cautiously if data fails QC steps or statistical method assumptions.</li>
<li>Identify deferentially expressed genes with raw sequencing counts (non-normalized data) using <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DeSeq2</a>. This can be done directly with DeSeq2 in R, or it can be done through HOMER using <a href="http://homer.ucsd.edu/homer/ngs/diffExpression.html"><code>getDiffExpression.pl</code></a>.</li>
<li>Results can be summarized genome wide with various types of scatter plots or heat maps.</li>
<li>Check biological function of differential genes using gene ontology tools. Examples include: <a href="https://metascape.org/gp/index.html">Metascape</a>, <a href="https://toppgene.cchmc.org">ToppGene</a> or <a href="https://toppgene.cchmc.org/enrichment.jsp">ToppFun</a>, <a href="https://www.gsea-msigdb.org/gsea/index.jsp">Gene Set Enrichment Analysis</a>, <a href="https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/">Qiagen Ingenuity Pathway Analysis</a> (<em>$$$</em>), etc.</li>
</ol>
</section>
<section id="analytical-creativity" class="level3">
<h3 class="anchored" data-anchor-id="analytical-creativity">Analytical creativity</h3>
<p>Analysis of RNA-seq data is only limited by the quality of the data and your imagination. Consider intersecting your results with data from other systems, experiments, or genomics assays as a way to develop and refine new hypotheses.</p>
</section>
</section>
<section id="automated-pipelines" class="level2">
<h2 class="anchored" data-anchor-id="automated-pipelines">Automated pipelines</h2>
<p>Applying the steps outlined above to a set of new data is cumbersome and prone to human typing errors. We have access to a few automated pipelines to help streamline these steps. However, it is a good idea to learn how to use the above tools in an independent workflow. And to learn how to apply the various options as best suited to your data type/assumptions. <em>Becoming over reliant on automated processing and analysis will limit understanding, independence, and creativity.</em></p>
<section id="map_on_bmi_hpc-in-progress" class="level3">
<h3 class="anchored" data-anchor-id="map_on_bmi_hpc-in-progress">map_on_bmi_hpc (in progress!)</h3>
<p><em>Presently, the lab has independent wrapper scripts individually tailored for single-read or paired-end RNA-seq data, or single-read or paired-end ChIP-seq or ATAC-seq data. Eventually these scripts will be merged into an integrated tool that functions via specification of command line options.</em></p>
<p>These wrapper scripts takes an input directory path of grouped data and generate and submit a series of batch .bat files on the BMI High Performance Cluster. Users have the option to specify variables specific to their data or use default specified parameters. The submitted batch files automate execution of the following: fastqc, trim_galore, data alignment, tag directory creation, and sorting and indexing alignment bam files.</p>
<p>These scripts can be run from the lab’s cluster drive or added as executables to your path. Current versions are located at:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/data/hpc-troutman/software/scripts/wrapper_star_paired.R</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">/data/hpc-troutman/software/scripts/working_wrapper_bowtie2_pairedend.R</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">/data/hpc-troutman/software/scripts/working_wrapper_bowtie2_singleread.R</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>These tools depend on file names in the lab convention (<code>*R1.fastq.gz *R2.fastq.gz</code>). Samples from the CCHMC core can be renamed as in the example below. <em>It is good practice to triple-check your commands with an <code>echo</code> statement before running the final command.</em></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>R1<span class="pp">*</span>.fastq.gz<span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> mv <span class="va">$i</span> <span class="va">${i</span><span class="op">/</span><span class="ss">R1_001.fastq.gz</span><span class="op">/</span>R1.fastq.gz<span class="va">}</span><span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>R2<span class="pp">*</span>.fastq.gz<span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> mv <span class="va">$i</span> <span class="va">${i</span><span class="op">/</span><span class="ss">R2_001.fastq.gz</span><span class="op">/</span>R2.fastq.gz<span class="va">}</span><span class="kw">;</span> <span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>These tools require input of only one <code>fastq.gz</code> file per sample per read direction.</p></li>
<li><p>These tools generate tag directories assuming a stranded library chemistry (e.g.&nbsp;lab standard <a href="https://cchmc-my.sharepoint.com/:w:/g/personal/ty_troutman_cchmc_org/EQq4IL-eReVMgL_-hyBey8IBvTnNd7zF6jL4pGzWQxyylw?e=uFlh8V">dUTP RNA-seq protocol</a>).</p></li>
</ul>
</div>
</div>
<p>Usage on the cluster requires a loaded R module. An example follows:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load R/4.2.1</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> /data/hpc-troutman/software/scripts/working_wrapper_star_paired.R <span class="at">-d</span> /data/troutman-hpc/archive/fastq_path</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For example, the script generates and submits the following .bat file to the LSF scheduler:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -W 2:00</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -n 16</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -M 40000</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -e ./bsub_scripts/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.err</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -o ./bsub_scripts/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.out</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#BSUB -R "span[hosts=1]"</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># load modules</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load fastqc/0.11.7</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load trimgalore/0.6.6</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load STAR/2.7.9</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load samtools</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load homer/4.11</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co"># execute fastqc</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> /data/hpc-troutman/data/archive/230608_CCHMC_SP/atf3//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R1.fastq.gz <span class="at">-o</span> ./fastqc_orig/</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> /data/hpc-troutman/data/archive/230608_CCHMC_SP/atf3//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R2.fastq.gz <span class="at">-o</span> ./fastqc_orig/</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="co"># execute trim_galore</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="dt">\</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">--cores</span> 4 <span class="dt">\</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">--paired</span> /data/hpc-troutman/data/archive/230608_CCHMC_SP/atf3//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R1.fastq.gz /data/hpc-troutman/data/archive/230608_CCHMC_SP/atf3//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R2.fastq.gz<span class="dt">\</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastqc</span> <span class="dt">\</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output_dir</span> ./trim_galore/</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a><span class="co"># rename trimmed files</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ./trim_galore/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R1_val_1.fq.gz ./trim_galore/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R1.fastq.gz</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ./trim_galore/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R2_val_2.fq.gz ./trim_galore/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R2.fastq.gz</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> ./trim_galore/<span class="pp">*</span>fastqc<span class="pp">*</span> ./fastqc_trimmed/</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a><span class="co"># execute STAR</span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="dt">\</span></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">--runThreadN</span> 16 <span class="dt">\</span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genomeDir</span> /data/hpc-troutman/data/genomes/indexes/star/mm10_starIndex/ <span class="dt">\</span></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesIn</span> ./trim_galore//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R1.fastq.gz ./trim_galore//Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> ./star_out/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10. <span class="dt">\</span></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a><span class="co"># remove STAR temporary directories</span></span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> ./star_out/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10._STARtmp</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a><span class="co"># execute HOMER makeTagDirectory</span></span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a><span class="ex">makeTagDirectory</span> ./tag_directories/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10/ <span class="dt">\</span></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>    ./star_out/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10.Aligned<span class="pp">*</span>am <span class="dt">\</span></span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">-genome</span> mm10 <span class="at">-checkGC</span> <span class="at">-format</span> sam <span class="at">-sspe</span> <span class="at">-flip</span></span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a><span class="co"># copy STAR log files to the HOMER tag directory</span></span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./star_out/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.<span class="pp">*</span>Log.final.out ./tag_directories/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.star.log.mm10</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ./star_out/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.<span class="pp">*</span>Log.out ./tag_directories/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.mm10/Mouse-C57BL6J-BMDM-RNA-ATF3ko-LPS-1hr-1407-m-E00457-TATTGAAT-TGCGTCGG.star.consolelog.mm10 <span class="dt">\</span></span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a> <span class="bu">echo</span> Done!</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="a.c.rsuite" class="level3">
<h3 class="anchored" data-anchor-id="a.c.rsuite"><a href="https://github.com/z5ouyang/A.C.Rsuite">A.C.Rsuite</a></h3>
<p>This analysis tool suite was created by Zhengyu Ouyang to simplify frequent HOMER analyses of mapped and processed genomics data (RNA-seq as well as ATAC-seq and ChIP-seq). These tools require an input <code>sample definition</code> file. This is a 4 column table for RNA-Seq or ATAC-Seq and 5 column table for ChIP-Seq. The table must be formatted as ‘tab’ separated (\t) and should not contain a header row.</p>
<ul>
<li>Each row in the table corresponds to a data group.</li>
<li>The first column is the name of the group.</li>
<li>The second column defines the color of the group. This can contain default <a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf">R colors</a> or hex formatted color codes (example: #de2d26).</li>
<li>The third column contains the paths to each HOMER tag directory belonging to the data group. Each sample/replicate directory for the group must be separated by a semicolon <code>;</code> (no space!).</li>
<li>The fourth column contains the name assigned to each sample. Each sample/replicate name for the group must be separated by a semicolon <code>;</code> (no space!). The order and number of sample names should be the same as the tag directories in column 3.</li>
</ul>
<p>An example follows (omit the header column):</p>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Column 1</th>
<th style="text-align: left;">Column 2</th>
<th style="text-align: left;">Column 3</th>
<th style="text-align: left;">Column 4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">group1</td>
<td style="text-align: left;">gray0</td>
<td style="text-align: left;">/path/to/tag_directory/group1_rep1/;/path/to/tag_directory/group1_rep2/</td>
<td style="text-align: left;">group1_rep1;group1_rep2</td>
</tr>
<tr class="even">
<td style="text-align: left;">group2</td>
<td style="text-align: left;">gray33</td>
<td style="text-align: left;">/path/to/tag_directory/group2_rep1/;/path/to/tag_directory/group2_rep2/</td>
<td style="text-align: left;">group2_rep1;group2_rep2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">group3</td>
<td style="text-align: left;">gray66</td>
<td style="text-align: left;">/path/to/tag_directory/group3_rep1/;/path/to/tag_directory/group3_rep2/</td>
<td style="text-align: left;">group3_rep1;group3_rep2</td>
</tr>
</tbody>
</table>
<p><em>The GitHub versions are not fully functional on the BMI cluster, but we have edited versions suppressing the non-functional aspects (<code>makeMultiWigHub.pl</code>).</em></p>
<p>####rnaPipe.R <code>rnaPipe.R</code> is a full featured wrapper script that executes the below three tools in series, generating a summary stats file called <code>alignStats.txt</code>, data quantification results in a directory called <code>rnaQuan/</code>, and differential expression results in a directory called <code>rnaDiff/</code>. These tools require a compatible R and HOMER module is loaded when run through the BMI HPC. An example usage follows:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load R/4.2.1</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load homer/4.11</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="ex">rnaPipe.R</span> /path/to/sample_definition_file <span class="at">-o</span> /desired/output/path <span class="at">-g</span> available_homer_genome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Steps from <code>rnaPipe.R</code> can also be run/rerun individually, and some have additional analysis options available. Script instructions and available options can be viewed by entering the command into the terminal.</p>
<section id="alignstats.r" class="level4">
<h4 class="anchored" data-anchor-id="alignstats.r">alignStats.R</h4>
<p><code>alignStats.R</code> compiles various quality control statistics into a summary table based on what is in the sample definition file. An example usage follows:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> alignStats.R /path/to/sample_definition_file <span class="op">&gt;</span> alignStats.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="rnaquan.r" class="level4">
<h4 class="anchored" data-anchor-id="rnaquan.r">rnaQuan.R</h4>
<p><code>rnaQuan.R</code> performs routine quantification of RNA-seq data by reading in the sample definition file and using HOMER and R to perform the following: - Generate an <code>analyzeRepeats.pl</code> raw sequencing count file (<code>-raw</code> or <code>-noadj</code>) specifying the options <code>-condenseGenes -count exons</code>. - Generate an <code>analyzeRepeats.pl</code> TPM normalized file (<code>-tpm</code> option) specifying the option <code>-count exons</code> <u>but not <code>-condenseGenes</code></u>. This file is subsetted based on accession numbers within the raw count file. - Short genes (less than 200 bp) are removed, and a simplified raw file (<code>rawC.txt</code>) and TPM file (<code>rawT.txt</code>) are created. - A series of plots are created that: compare gene length versus expression means and variance; assess data grouping using principal component analyses; compare pair-wise Parson’s correlations using heat maps and within group 2d kernel density plots. - Group replicate tag directories are merged into a <code>mergeTag</code> folder. - A UCSC genome browser hub track is created using the sample definition file name. <u>This aspect will not function until the web server and script are correctly configured</u>.</p>
<p>An example usage follows:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> rnaQuan.R /path/to/sample_definition_file <span class="at">-o</span> rnaQuan</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="rnadiff.r" class="level4">
<h4 class="anchored" data-anchor-id="rnadiff.r">rnaDiff.R</h4>
<p><code>rnaDiff.R</code> performs group pairwise differential expression analysis using DeSeq2 as. This depends both on the sample definition file <em>and</em> the processed raw count and TPM files generated by <code>rnaQuan.R</code>. It generates the following: - DeSeq2 results for any gene above a declared TPM threshold (default 8) for each pairwise comparison between each group (not one group versus all remaining). - Scatter plots and MA plots visualizing expression differences between each pair of groups. - A series of summary heat map for the composite list of deferentially expressed genes between all groups.</p>
<p>An example usage is follows:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> rnaDiff.R /path/to/sample_definition_file <span class="at">-c</span> rnaQuan/rawC.txt <span class="at">-t</span> rnaQuan/rawT.txt <span class="at">-o</span> rnaDiff/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>These scripts can be run from the lab’s cluster drive or added as an executable to your path. Current versions are located at <code>/data/hpc-troutman/software/scripts/</code>.</p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>